The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12807395249/job/35707903715) [(plain)](https://github.com/rust-lang/rust/commit/5951f9f12e5d3586ff799d59d5c6a9c11ce4eed5/checks/35707903715/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     |
1085 |     let sysroot = project_model::Sysroot::discover(
     |                   -------------------------------- arguments to this function are incorrect
...
1088 |         &project_model::SysrootQueryMetadata::default(),
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `SysrootQueryMetadata`, found `&SysrootQueryMetadata`
note: associated function defined here
    --> /checkout/src/tools/rust-analyzer/crates/project-model/src/sysroot.rs:126:12
     |
126  |     pub fn discover(
126  |     pub fn discover(
     |            ^^^^^^^^
help: consider removing the borrow
     |
1088 -         &project_model::SysrootQueryMetadata::default(),
1088 +         project_model::SysrootQueryMetadata::default(),

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rust-analyzer` (test "slow-tests") due to 1 previous error
Build completed unsuccessfully in 0:05:16
  local time: Thu Jan 16 11:14:35 UTC 2025
  network time: Thu, 16 Jan 2025 11:14:35 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/src/tools/rust-analyzer/crates/proc-macro-srv/src/lib.rs
+++ b/src/tools/rust-analyzer/crates/proc-macro-srv/src/lib.rs
@@ -175,7 +175,7 @@ fn expand_id(
             });
         let res = match thread {
             Ok(handle) => handle.join(),
-            Err(e) => std::panic::resume_unwind(Box::new(e)),
+            Err(e) => return Err(e.to_string()),
         };
 
         match res {
@@ -223,7 +223,7 @@ fn expand_ra_span(
             });
         let res = match thread {
             Ok(handle) => handle.join(),
-            Err(e) => std::panic::resume_unwind(Box::new(e)),
+            Err(e) => return Err(e.to_string()),
         };
 
         match res {

--- a/src/tools/rust-analyzer/crates/ra-salsa/src/lib.rs
+++ b/src/tools/rust-analyzer/crates/ra-salsa/src/lib.rs
@@ -610,11 +610,9 @@ where
 #[non_exhaustive]
 pub enum Cancelled {
     /// The query was operating on revision R, but there is a pending write to move to revision R+1.
-    #[non_exhaustive]
     PendingWrite,
 
     /// The query was blocked on another thread, and that thread panicked.
-    #[non_exhaustive]
     PropagatedPanic,
 }
 

--- a/src/tools/rust-analyzer/crates/rust-analyzer/src/handlers/dispatch.rs
+++ b/src/tools/rust-analyzer/crates/rust-analyzer/src/handlers/dispatch.rs
@@ -5,6 +5,7 @@ use std::{
 };
 
 use ide::Cancelled;
+use ide_db::base_db::ra_salsa::Cycle;
 use lsp_server::{ExtractError, Response, ResponseError};
 use serde::{de::DeserializeOwned, Serialize};
 use stdx::thread::ThreadIntent;
@@ -328,7 +329,13 @@ where
             if let Some(panic_message) = panic_message {
                 message.push_str(": ");
                 message.push_str(panic_message)
-            };
+            } else if let Some(cycle) = panic.downcast_ref::<Cycle>() {
+                tracing::error!("Cycle propagated out of salsa! This is a bug: {cycle:?}");
+                return Err(Cancelled::PropagatedPanic);
+            } else if let Ok(cancelled) = panic.downcast::<Cancelled>() {
+                tracing::error!("Cancellation propagated out of salsa! This is a bug");
+                return Err(*cancelled);
+            }
 
             Ok(lsp_server::Response::new_err(
                 id,

